pkg/sql/sem/tree/type_check.go

```diff
@@ -2588,8 +2597,17 @@ func typeCheckSameTypedExprs(
 	}
 
 	// Handle tuples, which will in turn call into this function recursively for each element.
-	if _, ok := exprs[0].(*Tuple); ok {
-		return typeCheckSameTypedTupleExprs(ctx, semaCtx, desired, exprs...)
+	if idx, ok := findFirstTupleIndex(exprs...); ok {
+		if _, ok := exprs[idx].(*Tuple); ok {
+			// typeCheckSameTypedTupleExprs expects the first expression in the slice
+			// to be a tuple.
+			exprs[0], exprs[idx] = exprs[idx], exprs[0]
```

Comment:
I'm not convinced it is safe to reorder subexpressions. Won't this change the semantics of parent nodes that pass in their `exprs` directly from an AST node, like `CoalesceExpr.TypeCheck`: https://github.com/cockroachdb/cockroach/blob/1758e91c7fe48dd3032bf9479402000eed393146/pkg/sql/sem/tree/type_check.go#L888